Add tests for Watchdog HAL API

fkjagodzinski opened the pull request:
### Description
Add tests for new Watchdog API introduced in #5488:
* `tests-mbed_hal-watchdog`,
* `tests-mbed_hal-watchdog_reset`,
* `tests-mbed_hal-watchdog_timing`,
* `tests-mbed_drivers-watchdog`,
* `tests-mbed_drivers-watchdog_reset`.

CC @bulislaw @0xc0170 @scartmell-arm

fkjagodzinski commented on TESTS/mbed_hal/watchdog_reset/watchdog_reset_tests.h:
```diff
@@ -0,0 +1,79 @@
+ * When the watchdog is started again and it's timeout expires.
+ * Then the device is restarted.
+ */
+void test_watchdog_operation();
```
Since all the logic is provided by the host side all these test cases are handled same way on the dev side -- by simply handling the commands. I feel they should get separated anyway to be consistent with other test suits. What would be the best way to do that? Copy the code into separate test suits by e.g. `TESTS/mbed_hal/{watchdog_reset,watchdog_kick,watchdog_sleep,watchdog_deepsleep,watchdog_disable}`? This would introduce code duplication on dev and host side.

bulislaw commented:
@jamesbeyond @scartmell-arm

0xc0170 commented:
I fast forwarded feature-watchdog branch to get latest travis fix , that should fix it also here, can you rebase please?

fkjagodzinski commented:
Rebased onto updated `ARMmbed:feature-watchdog`.

adbridge commented:
@0xc0170 can you take another look after the rebase please.

jamesbeyond commented on TESTS/mbed_hal/watchdog/main.cpp:
```diff
@@ -0,0 +1,125 @@
+
+    config.timeout_ms = features.max_timeout - 1;
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_init(&config));
+    TEST_ASSERT_EQUAL_UINT32(features.max_timeout - 1, hal_watchdog_get_reload_value());
```
After K64F watchdog stop problem got fixed 2e7ce31a9a834209e6ac7a4fa9a0ec8d9672f246 , this test still failed here on K64F, and got an error of:
```[1515493472.85][CONN][RXD] >>> Running case #4: 'Update config with multiple init calls'...
[1515493472.91][CONN][INF] found KV pair in stream: {{__testcase_start;Update config with multiple init calls}}, queued...
[1515493472.96][CONN][RXD] :83::FAIL: Expected 4294967287 Was 500
[1515493473.03][CONN][INF] found KV pair in stream: {{__testcase_finish;Update config with multiple init calls;0;1}}, queued...
[1515493473.12][CONN][RXD] >>> 'Update config with multiple init calls': 0 passed, 1 failed with reason 'Assertion Failed'```

scartmell-arm replied:
I'll look into it. But, I think it would be best if this function is removed. I can't see it being very useful as the user should know what timeout they set, and it may not return the same value as it's recalculating the timeout from the register*prescaler which may be slightly different to what they initially set.

fkjagodzinski replied:
Doesn't K64F have some restriction on R/W access to watchdog control registers? Adding `wait_us(1);` in between `hal_watchdog_init(&config)` calls seem to help.

scartmell-arm replied:
That's most likely the cause, adding the wait is a reasonable solution.

fkjagodzinski replied:
I think this should rather be buried inside a platform specific implementation. If K64F needs to do some waiting, then `hal_watchdog_init()` should not return until everything is ready. Alternatively, `hal_watchdog_init()` could perform some checks before doing the actual init. I asked @0xc0170 for a comment and he has a similar opinion.

fkjagodzinski commented:
Added `tests-mbed_hal-watchdog_timing` tests that check watchdog timer time accuracy with the help of a dedicated host script.

0xc0170 commented:
/morph build

0xc0170 commented:
Based on the test results, this was known issue that was resolved few days ago. I rebased the destination branch

fkjagodzinski commented:
Added tests for new driver APIs.

0xc0170 commented:
@fkjagodzinski Watchdog landed to feature branch, this should be rebased now.

fkjagodzinski commented:
Rebase onto `ARMmbed:feature-watchdog` done.

0xc0170 commented:
/morph build

0xc0170 commented:
The fix did not propagate, I pushed an update to feature-watchdog just now to fix the build failure

fkjagodzinski commented:
@0xc0170 Ok, rebased again.

cmonr commented:
/morph build

0xc0170 commented:
Test failures for some targets in the new test cases, please verify

fkjagodzinski commented:
ad7c530592aac1d36917e4ab03097e1e8ee0104f should fix all test cases that had an `ERROR` status due to undefined `'program_cycle_s'` config property.

jamesbeyond commented:
I believe @studavekar is fixing this bug in greentea, will create a pull request soon

cmonr commented:
@jamesbeyond Thanks for the heads up about the incoming PR. If you or @studavekar could link it in this PR when it's ready, that would be incredibly useful.

Unfortunately, this means that this PR won't be able to move forward until the GreenTea PR is not only merged, but released with the change, since our CI only uses released versions of the tools.

studavekar commented:
Have created a PR for greentea 

> Unfortunately, this means that this PR won't be able to move forward until the GreenTea PR is not only merged, but released with the change, since our CI only uses released versions of the tools.

Have created a PR for greentea https://github.com/ARMmbed/greentea/pull/261 , we work off master to once this is merged we should be good.

On side not when the tests are run using remote device it still fails with assert 
```
[1517592872.48][CONN][RXD] >>> Running case #1: 'Timing, 200 ms'...
[1517592872.48][CONN][INF] found KV pair in stream: {{__testcase_start;Timing, 200 ms}}, queued...
[1517592872.58][CONN][RXD] :50::FAIL: Values Not Within Delta 40 Expected 200 Was 86
[1517592872.66][CONN][INF] found KV pair in stream: {{__testcase_finish;Timing, 200 ms;0;1}}, queued.
```

0xc0170 commented:
@fkjagodzinski Have you made any progress with the failures?

fkjagodzinski commented:
@0xc0170 Yeah, I'm about to push fixes for stm32 in a moment.

fkjagodzinski commented:
92be47fac303f9c8f2ed4383e74613050e150bf0 should fix majority of issues with STM32F4 found during tests.

Depending on the timeout value, the resolution of the watchdog timer can go as low as 8 ms for STM32F4 so I'll also have to update the test code.

cmonr commented:
/morph build

0xc0170 commented:
@fkjagodzinski Please review tests results

fkjagodzinski commented:
91722304b078bead2a3502263e122785a9193539 fixes:
* `tests-mbed_drivers-watchdog` / `Restart multiple times`,
* `tests-mbed_hal-watchdog` / `Update config with multiple init calls`

for STM32F4 boards.

These cases still fail on K64F due to [this issue](https://github.com/ARMmbed/mbed-os/pull/5751#pullrequestreview-87477557) which hasn't been resolved yet.

Regarding `tests-mbed_hal-watchdog_timing`, I think gathering timestamps on dev side instead of the host side should help; will push that patch shortly.

fkjagodzinski commented:
a2c349fc975524979853a21745dda464b9df23fe will fix `tests-mbed_hal-watchdog_timing`. I suppose the issue was caused by a large/variable dev-host delay present during CI testruns.

fkjagodzinski commented:
@0xc0170 @cmonr @kegilbert Could you rerun the morph tests, please? There are still some things to sort out, but I'd like to see how those latest patches work on the CI.

cmonr commented:
/morph build

fkjagodzinski commented:
Fixed ARM and IAR build errors.

bulislaw commented:
/morph build

fkjagodzinski commented:
Update:
* extracted Reset Reason tests to a separate PR: #6269,
* squashed commits to improve readability,
* updated `tests-mbed_hal-watchdog_reset` -- moved the test logic to the device side to minimize the possibility of dev-host UART errors,
* updated timing tolerance in `tests-mbed_hal-watchdog_timing`,
* added a fix for K64F `hal_watchdog_init()` and `hal_watchdog_stop()`.

cmonr commented:
/morph build

studavekar commented:
re-trigger exporter, it failed because connection dropping between AWS and Jenkins. 

/morph export-build

fkjagodzinski commented:
```
Error: L6200E: Symbol __asm___9_em_vcmp_c_9440395e____REV16 multiply defined (by mbed-os/targets/TARGET_Silicon_Labs/TARGET_EFM32/emlib/src/em_vcmp.o and mbed-os/features/mbedtls/src/debug.o).
Error: L6200E: Symbol __asm___9_em_vcmp_c_9440395e____REVSH multiply defined (by mbed-os/targets/TARGET_Silicon_Labs/TARGET_EFM32/emlib/src/em_vcmp.o and mbed-os/features/mbedtls/src/debug.o).
Error: L6200E: Symbol __asm___9_em_vcmp_c_9440395e____RRX multiply defined (by mbed-os/targets/TARGET_Silicon_Labs/TARGET_EFM32/emlib/src/em_vcmp.o and mbed-os/features/mbedtls/src/debug.o).
Finished: 0 information, 0 warning and 3 error messages.
make[1]: *** [mbed-os-example-blinky.elf] Error 1
make: *** [all] Error 2
FAILURE
FAILURE building mbed-os-example-blinky EFM32WG_STK3800 make_armc5
```

`ci-morph-exporter` failures seem unrelated to changes introduced in this PR.

cmonr commented:
Going to try and restart the build. @kegilbert @studavekar wondering if this is similar to an error we saw earlier today.

/morph export-build
